### ISSUE TITLE
perf(model): keep immutable CUDA weights resident across exact cache hits

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#9420]
## Problem frame
Centrality: Core.

For operators serving Qwen3.8 through the fak-native CUDA path; the problem is that exact radix-cache hits enter decode with a fresh session and restage immutable model weights; today cold prefill warms session-local weights but a later exact hit skips prefill, so its first decode step absorbs the cold upload; this is better because model-lifetime immutable residency removes repeated setup without changing KV/request ownership or silently delegating execution; the witness is one-upload-across-sessions ownership tests plus a matched A100 cache-hit receipt naming `engine=fak-native` and the exact Qwen3.8 artifact.

- P1 Context: advanced — exact prefix reuse must remove repeated setup rather than move it into decode.
- P2 Net value: advanced — the committed A100 attribution measures device-prefix clone at about 29 ms while hit decode is about 4.7 s; require net matched-envelope improvement with quality held.
- P3 Adaptation: advanced — keep residency bounded to immutable model weights, one backend identity, explicit model teardown, and fail closed on incompatible backend reuse.
- P4 Operations: advanced — preserve cache/engine identity in receipts and prove lifecycle behavior locally before the sanctioned A100 run.

## Reproduction
`InKernelPlanner` creates a new backend session for a device-prefix hit. `Session.halW` starts empty. Exact cached logits bypass prefill. The first `Step` lazily stages weights, while `Session.Close` frees all session-owned `halW`, so the next request repeats the work.

The issue-8819 A100 roofline witness records four device-clone events of roughly 28–29 ms for 161,218,560 bytes, versus roughly 4.7 s of cache-hit decode. This selects immutable device-weight lifetime as the next setup bottleneck; it does not claim host L2/L3 restore is fast.

## Smallest spine
- Add model-lifetime, concurrency-safe immutable device-weight residency for device backends.
- Sessions borrow resident immutable weights and continue owning request-local KV, recurrent state, activations, and non-shared tensors.
- Repeated sequential and concurrent sessions upload a given immutable weight once.
- Closing one session cannot invalidate another borrower; model weight teardown frees residency exactly once.
- Preserve CPU behavior and reject unsafe backend-identity changes.

## Witness
- Failure-first tests prove one upload across two sessions, first-session close does not invalidate the second, and model teardown frees once.
- Existing model/agent tests remain green through the scope-correct isolated gate.
- A matched 5-cold + 5-identical-hit A100 receipt uses Qwen3.8-27B Q4_K_M, exact artifact hash, `engine=fak-native`, zero fallback, exact output, cache-source/cached-token evidence, weight-upload bytes, clone timing, and first-step/steady-decode timing.

## Gold-plating boundary
Do not combine paged-swap serialization, warm-splice relaunch, host L2/L3, a CUDA kernel rewrite, or session pooling into this leaf. File remaining measured bottlenecks separately.

Depends on #8819. Related but distinct from CUDA-kernel lever #8821 and placement policy #9357.

**Comment by anthony-chaudhary** [comment]
Shipped the local correctness spine in `internal/model@r507+ged922ce94` (`ed922ce94b38e0f3d0892303385635f3f5cf6d3a`). The commit is on `origin/main` and `dos commit-audit` classifies it `diff-witnessed`.

What changed:
- model-lifetime, backend-identity-scoped residency now owns immutable CUDA weights;
- sessions borrow resident handles while retaining request-local KV/recurrent/activation ownership;
- routed experts outside the bounded expert ring stay session-owned;
- teardown frees shared handles once and preserves the existing Q4K/Metal close order;
- fail-closed token-lineage teardown now captures the admitted KV store, avoiding a nil dereference after backend closure.

Post-commit witness:

```text
FAK_FAST=0; .\test.ps1 ./internal/model -count=1
ok github.com/anthony-chaudhary/fak/internal/model 24.264s
```

This is `CLAIM_TEST_GREEN`. The eight-path commit includes failure-first sequential/concurrent one-upload tests, backend isolation, Q4K/Qwen-derived residency, CPU preservation, O(1) borrowed-key close behavior, and exact-once model teardown.

Commit-time prospective buildcheck exceeded its 2-minute host deadline and was explicitly recorded fail-open; the independent full affected-package WSL run above is the completion evidence, not that timeout.

Hardware performance status remains **PENDING_AUTH**, so this comment does not claim the A100 throughput is fixed. Both sanctioned routes were exercised: the private bridge readiness check lacks its operator credential/channel, and the GCP probe cannot refresh its stale credential noninteractively. The pinned A100 public-cloud plan is valid in dry-run:

```powershell
gcloud auth login
python tools/gcp_gpu_probe.py --all-tiers
python tools/gcp_bench.py --tier a2-high-a100-40gb-1g --engine fak-cuda --fak-ref ed922ce94b38e0f3d0892303385635f3f5cf6d3a --keep
```

On that node, rerun the issue witness exactly as scoped: same-process 5 cold + 5 identical exact-output requests against Qwen3.8-27B Q4_K_M artifact SHA-256 `7e78…e169`, `engine=fak-native`, zero fallback, request-bound cache tier/cached tokens, immutable-weight upload bytes/calls, clone time, cached-logit TTFT, first real `Step`, steady decode, and 5/5 exact output. Acceptance is zero repeated immutable-weight upload on hits plus a matched before/after latency receipt. Keep this issue open until that receipt lands.

**Comment by anthony-chaudhary** [comment]
Resolved by `ed922ce` (resolving commit). Closed by the DOS dispatch loop's close-resolved arm, witnessed via `dos commit-audit` (verdict OK / diff-witnessed). Reopen if this does not fully resolve it.

**Comment by anthony-chaudhary** [comment]
Reopened by the two-hour completion audit: the issue's own shipped-spine comment records PENDING_AUTH and explicitly requires the matched A100 cache-hit receipt before closure. A diff witness is not the required hardware/performance witness.

**Comment by anthony-chaudhary** [comment]
Instrumentation needed for the hardware closure receipt is now on origin/main: #9794 plus hardening commit 504bf5a0c (internal/agent@r366+g504bf5a0c; internal/compute@r216+g504bf5a0c). It exposes cumulative CUDA immutable-weight upload before/after/delta and structured real-forward timing, with cache-hit, rank-one, Q2 ownership, and Device-L1 exact-hit witnesses. #9420 remains open for the exact fak-native A100 5-cold/5-hit campaign; no local-host substitute or fallback claim is being made.

**Comment by anthony-chaudhary** [comment]
Closing as shipped on origin/main in commit ed922ce94b38e0f3d0892303385635f3f5cf6d3a ($sub). The commit directly implements the issue-named behavior and includes its repository witness; current trunk contains it.
